File: MinCost.java
/*this implementation uses dynamic programming approach to determine the minimum cost to paint houses
 * the time complexity of this approach is O(n)*/
class MinCost {
    public int minCost(int[][] costs) {
        if (costs == null || costs.length == 0) {
            return 0;
        }
        int n = costs.length;
        int[] prevRow = new int[3];
        prevRow[0] = costs[0][0];
        prevRow[1] = costs[0][1];
        prevRow[2] = costs[0][2];
        int prevMin = Math.min(Math.min(prevRow[0], prevRow[1]), prevRow[2]);
        for (int i = 1; i < n; i++) {
            int[] curRow = new int[3];
            curRow[0] = costs[i][0] + Math.min(prevRow[1], prevRow[2]);
            curRow[1] = costs[i][1] + Math.min(prevRow[0], prevRow[2]);
            curRow[2] = costs[i][2] + Math.min(prevRow[0], prevRow[1]);
            int curMin = Math.min(Math.min(curRow[0], curRow[1]), curRow[2]);
            prevRow = curRow;
            prevMin = curMin;
        }
        return prevMin;
    }
    public static void main(String[] args) {
        int[][] costs = {{17, 2, 17}, {16, 16, 5}, {14, 3, 19}};
        MinCost solution = new MinCost();
        int minCost = solution.minCost(costs);
        System.out.println("Minimum cost to paint all houses: " + minCost);
    }
}

File: minCost.py
#Time Complexity=O(m*3)=O(m),m-number of houses n=3-length of the array having cost for each house to paint
#Space Complexity=O([m+1]*3)=O(m)=m=length of the houses
#Ran successfully in Leetcode:Yes
#Here we first take alternate indices and calculate the min vost and store in an array. Then from each row of the array we choose min value and add them up to give solution.
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        if not costs:
            return 0
        if len(costs)==0:
            return 0
        n=len(costs)-1
        for i in range(1,len(costs)):
            costs[i][0]+=min(costs[i-1][2],costs[i-1][1])
            costs[i][1]+=min(costs[i-1][0],costs[i-1][2])
            costs[i][2]+=min(costs[i-1][0],costs[i-1][1])
        return min(costs[n][0],costs[n][1],costs[n][2])
        
